src/librustc_parse/parser/expr.rs

```diff
@@ -998,6 +1008,21 @@ impl<'a> Parser<'a> {
         }
     }
 
+    fn maybe_collect_tokens(
+        &mut self,
+        has_outer_attrs: bool,
+        f: impl FnOnce(&mut Self) -> PResult<'a, P<Expr>>,
+    ) -> PResult<'a, P<Expr>> {
+        if has_outer_attrs || self.is_collecting_tokens() {
```

Comment:
```suggestion
        if has_outer_attrs {
```
?

When collecting tokens for `#[attr] f(g)` we only need to assign a `TokenSteam` to `f(g)`, but not to `g`, because `g` doesn't have any attributes applied to it.
This should eliminate the need in the `collecting_tokens` field as well.